sigmastate/src/main/scala/sigmastate/interpreter/CostAccumulator.scala

```diff
@@ -0,0 +1,75 @@
+package sigmastate.interpreter
+
+import sigmastate.lang.exceptions.CostLimitException
+
+/** Encapsulate simple monotonic (add only) counter with reset. */
+class CostCounter(val initialCost: Int) {
+  private var _currentCost: Int = initialCost
+
+  @inline def += (n: Int) = {
+    this._currentCost = java.lang.Math.addExact(this._currentCost, n)
+  }
+  @inline def currentCost: Int = _currentCost
+  @inline def resetCost() = { _currentCost = initialCost }
+}
+
+/** Implements finite state machine with stack of graph blocks (scopes),
+  * which correspond to lambdas and thunks.
+  * It accepts messages: startScope(), endScope(), add(), reset()
+  * At any time `totalCost` is the currently accumulated cost. */
+class CostAccumulator(initialCost: Int, costLimit: Option[Long]) {
+
+  @inline private def initialStack() = List(new Scope(initialCost))
+  private var _scopeStack: List[Scope] = initialStack
+
+  @inline def currentScope: Scope = _scopeStack.head
+
+  /** Represents a single scope during execution of the graph.
+    * The lifetime of each instance is bound to scope execution.
+    * When the evaluation enters a new scope (e.g. calling a lambda) a new Scope instance is created and pushed
+    * to _scopeStack, then is starts receiving `add` method calls.
+    * When the evaluation leaves the scope, the top is popped off the stack. */
+  class Scope(initialCost: Int) extends CostCounter(initialCost) {
+
+
+    @inline def add(opCost: Int): Unit = {
+          this += opCost
+    }
+
+    /** Called by nested Scopes to communicate accumulated cost back to parent scope.
+      * When current scope terminates, it communicates accumulated cost up to its parent scope.
+      * This value is used at the root scope to obtain total accumulated scope.
+      */
+    private var _resultRegister: Int = 0
+    @inline def childScopeResult: Int = _resultRegister
+    @inline def childScopeResult_=(resultCost: Int): Unit = {
+      _resultRegister = resultCost
+    }
+
+  }
+
+  /** Called once for each operation of a scope (lambda or thunk).
+    */
+  def add(opCost: Int): Unit = {
```

Comment:
Please mark that the method is possibly throwing an exception somehow

Reply:
done